src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildAgentSSHKeyService.java

```diff
@@ -0,0 +1,110 @@
+package de.tum.in.www1.artemis.service.connectors.localci.buildagent;
+
+import static de.tum.in.www1.artemis.config.Constants.PROFILE_BUILDAGENT;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.security.GeneralSecurityException;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.util.Optional;
+
+import org.apache.sshd.common.config.keys.writer.openssh.OpenSSHKeyEncryptionContext;
+import org.apache.sshd.common.config.keys.writer.openssh.OpenSSHKeyPairResourceWriter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.context.annotation.Profile;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Service;
+
+@Service
+@Profile(PROFILE_BUILDAGENT)
+public class BuildAgentSSHKeyService {
+
+    private static final Logger log = LoggerFactory.getLogger(BuildAgentSSHKeyService.class);
+
+    private KeyPair keyPair;
+
+    @Value("${artemis.version-control.ssh-private-key-folder-path:#{null}}")
+    private Optional<String> gitSshPrivateKeyPath;
+
+    @Value("${artemis.version-control.build-agent-use-ssh:false}")
+    private boolean useSSHForBuildAgent;
+
+    @Value("${info.contact}")
+    private String sshKeyComment;
+
+    /**
+     * Generates the SSH key pair and writes the private key when the application is started and the build agents should use SSH for their git operations.
+     */
+    @EventListener(ApplicationReadyEvent.class)
+    public void applicationReady() {
+        if (!useSSHForBuildAgent) {
+            return;
+        }
+
+        log.info("Using SSH for build agent authentication.");
+
+        if (gitSshPrivateKeyPath.isEmpty()) {
+            throw new RuntimeException("No SSH private key folder was set but should use SSH for build agent authentication.");
+        }
+
+        generateKeyPair();
+
+        try {
+            writePrivateKey();
+        }
+        catch (IOException | GeneralSecurityException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void generateKeyPair() {
+        try {
+            KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
+            keyGen.initialize(4096);
+            keyPair = keyGen.generateKeyPair();
+        }
+        catch (GeneralSecurityException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
Optional:
As this method is only used in the `applicationReady` method together with the `writePrivateKey` method, which already throws a `GeneralSecurityException`, you could simple remove the catch here, declare the exception for this method and move it into the existing try-catch block in `applicationReady`.